src/hotspot/share/opto/superword.cpp

```diff
@@ -1352,59 +1349,43 @@ int SuperWord::est_savings(Node* s1, Node* s2) {
   return MAX2(save_in, save_use);
 }
 
-//------------------------------costs---------------------------
-int SuperWord::adjacent_profit(Node* s1, Node* s2) { return 2; }
-int SuperWord::pack_cost(int ct)   { return ct; }
-int SuperWord::unpack_cost(int ct) { return ct; }
-
-// Combine packs A and B with A.last == B.first into A.first..,A.last,B.second,..B.last
+// Combine pairs (n1, n2), (n2, n3), ... into pack (n1, n2, n3 ...)
 void SuperWord::combine_pairs_to_longer_packs() {
 #ifdef ASSERT
-  assert(!_packset.is_empty(), "packset not empty");
-  for (int i = 0; i < _packset.length(); i++) {
-    assert(_packset.at(i) != nullptr, "no nullptr in packset");
-    assert(_packset.at(i)->size() == 2, "all packs are pairs");
-  }
+  assert(!_pairset.is_empty(), "pairset not empty");
+  assert(_packset.is_empty(), "packset not empty");
 #endif
 
-  bool changed = true;
-  // Combine packs regardless max vector size.
-  while (changed) {
-    changed = false;
-    for (int i = 0; i < _packset.length(); i++) {
-      Node_List* p1 = _packset.at(i);
-      if (p1 == nullptr) continue;
-      // Because of sorting we can start at i + 1
-      for (int j = i + 1; j < _packset.length(); j++) {
-        Node_List* p2 = _packset.at(j);
-        if (p2 == nullptr) continue;
-        if (p1->at(p1->size()-1) == p2->at(0)) {
-          for (uint k = 1; k < p2->size(); k++) {
-            p1->push(p2->at(k));
-          }
-          _packset.at_put(j, nullptr);
-          changed = true;
-        }
-      }
+  // Iterate pair-chain by pair-chain, each from left-most to right-most.
+  Node_List* pack = nullptr;
+  for (PairSetIterator pair(_pairset); !pair.done(); pair.next()) {
+    Node* s1 = pair.left();
+    Node* s2 = pair.right();
+    if (_pairset.is_left_in_a_left_most_pair(s1)) {
+      pack = new (arena()) Node_List(arena());
+      pack->push(s1);
+    }
+    pack->push(s2);
```

Comment:
We should also assert that at this point, `pack != nullptr`.

Reply:
Can do that, but I think it would just result in a nullptr exception anyway.
I always wonder if an assert makes the code more cluttered, so less readable, or if it states expectations more explicitly, which makes the code more readable :sweat_smile:

Reply:
Right, well then you can leave it as is. One sometimes tends to be paranoid when it comes to using asserts in C2 :-)